File: server/src/services/model-sync/scheduler.ts
import { syncModels } from './sync.js';

const SYNC_INTERVAL_MS = 6 * 60 * 60 * 1000;

let intervalId: ReturnType<typeof setInterval> | null = null;

export async function runInitialSync(): Promise<void> {
  console.log('[ModelSync] Running initial sync...');
  try {
    const result = await syncModels();
    console.log(
      `[ModelSync] Initial sync done: ${result.added} added (active), ${result.storedDisabled} stored (disabled), ` +
      `${result.updated} updated, ${result.freeToPaid} free→paid, ${result.paidToFree} paid→free`
    );
  } catch (err) {
    console.error('[ModelSync] Initial sync failed:', err);
  }
}

export function startSyncScheduler(): void {
  if (intervalId) return;
  console.log(`[ModelSync] Scheduler started (every ${SYNC_INTERVAL_MS / 3600000}h)`);

  intervalId = setInterval(async () => {
    console.log('[ModelSync] Running scheduled sync...');
    try {
      const result = await syncModels();
      console.log(
        `[ModelSync] Sync done: ${result.added} added, ${result.updated} updated, ${result.disabled} disabled`
      );
    } catch (err) {
      console.error('[ModelSync] Scheduled sync failed:', err);
    }
  }, SYNC_INTERVAL_MS);
}

export function stopSyncScheduler(): void {
  if (intervalId) {
    clearInterval(intervalId);
    intervalId = null;
  }
}

File: crates/server/src/services/model_sync/scheduler.rs
//! Port of `server/src/services/model-sync/scheduler.ts`.

use std::sync::atomic::{AtomicBool, Ordering};

use super::sync::sync_models;

const SYNC_INTERVAL_MS: u64 = 6 * 60 * 60 * 1000;

pub async fn run_initial_sync() {
    tracing::info!("[ModelSync] Running initial sync...");
    match sync_models().await {
        Ok(result) => {
            tracing::info!(
                "[ModelSync] Initial sync done: {} added (active), {} stored (disabled), \
                 {} updated, {} free→paid, {} paid→free",
                result.added,
                result.stored_disabled,
                result.updated,
                result.free_to_paid,
                result.paid_to_free
            );
        }
        Err(err) => {
            tracing::error!("[ModelSync] Initial sync failed: {err}");
        }
    }
}

pub fn start_sync_scheduler() {
    static STARTED: AtomicBool = AtomicBool::new(false);
    if STARTED.swap(true, Ordering::SeqCst) {
        return;
    }
    tracing::info!(
        "[ModelSync] Scheduler started (every {}h)",
        SYNC_INTERVAL_MS / 3600000
    );

    tokio::spawn(async move {
        let mut ticker = tokio::time::interval(std::time::Duration::from_millis(SYNC_INTERVAL_MS));
        ticker.tick().await; // skip immediate first tick (setInterval semantics)
        loop {
            ticker.tick().await;
            tracing::info!("[ModelSync] Running scheduled sync...");
            match sync_models().await {
                Ok(result) => {
                    tracing::info!(
                        "[ModelSync] Sync done: {} added, {} updated, {} disabled",
                        result.added,
                        result.updated,
                        result.disabled
                    );
                }
                Err(err) => {
                    tracing::error!("[ModelSync] Scheduled sync failed: {err}");
                }
            }
        }
    });
}
